### PR TITLE
Add option to backup and restore snapshots

### PR DESCRIPTION
**Pull request by fsperling**
- works in the same way as all the other grafana api entities
- i also fixed a few linter errors/warnings

### CODE REVIEW

**Comment by acjohnson** [comment]
@fsperling I'm hoping to get this merged but was wondering why `grafana_backup/clean_grafana.py` appears to be a copy of `cli.py` and why not just combine your `delete` command into the main cli parser in that file instead? Thanks!

**Comment by fsperling** [comment]
@acjohnson Indeed it would be easier to add the delete command to the cli.py.
I didn't do it like that because while it can be a useful feature it goes a bit beyond backup/restore functionality.
Currently I'm working on a command to pause/unpause certain alerts. I'm also planning to write another that changes users rights to viewers/editors. Things I plan to use in an upcoming migration.
All of this goes a bit into a grafana-utils direction so I'm a bit hesitant to just add it to all tool called grafana-backup.

**Comment by acjohnson** [comment]
@fsperling I don't like the idea of adding bespoke arg parsers to this project, personally I think the `delete` command is perfectly acceptable to be added to `cli.py` maybe @ysde or others will have a different take.

**Comment by fsperling** [comment]
@acjohnson I'm fine with that. I've updated the PR so that the delete command is part of the cli.py.

File: grafana_backup/dashboardApi.py [hunk 1]
```diff
@@ -92,8 +112,28 @@ def create_alert_channel(payload, grafana_url, http_post_headers, verify_ssl, cl
...
+
+
+def delete_snapshot(key, grafana_url, http_post_headers):
+    r = requests.delete('{0}/api/snapshots/{1}'.format(grafana_url, key), headers=http_post_headers)
+    return int(r.status_code)
+
+
 def delete_dashboard(board_uri, grafana_url, http_post_headers):
-    r = requests.delete('{0}/api/dashboards/db/{1}'.format(grafana_url, board_uri), headers=http_post_headers)
+    r = requests.delete('{0}/api/dashboards/uid/{1}'.format(grafana_url, board_uri), headers=http_post_headers)
```
Comment (acjohnson) [inline grafana_backup/dashboardApi.py:136]: I'm pretty sure this change will cause breakage in older versions of grafana @fsperling . You could probably use `uid_feature_check` to determine if `uid` is available but from what I recall if we change this here the `delete_dashboard` function will stop working on grafana versions earlier than `5.0.0` or something

File: grafana_backup/dashboardApi.py [hunk 2]
```diff
@@ -107,11 +147,32 @@ def search_datasource(grafana_url, http_get_headers, verify_ssl, client_cert, de
...
+                             client_cert, debug)
+
+
 def create_datasource(payload, grafana_url, http_post_headers, verify_ssl, client_cert, debug):
     return send_grafana_post('{0}/api/datasources'.format(grafana_url), payload, http_post_headers, verify_ssl,
                              client_cert, debug)
 
 
+def delete_datasource(uid, grafana_url, http_post_headers, verify_ssl, client_cert, debug):
+    r = requests.delete('{0}/api/datasources/uid/{1}'.format(grafana_url, uid), headers=http_post_headers)
```
Comment (acjohnson) [inline grafana_backup/dashboardApi.py:172]: same thing here, if we want to maintain backwards compat, we should use `uid_feature_check` to ensure older versions of grafana keep working with all features of this tool, even new features if/when possible.

**Comment by fsperling** [comment]
@acjohnson I updated the PR by adding support for older versions of Grafana
you can see the changes best here:
https://github.com/idealo/grafana-backup-tool/pull/5/files

**Comment by acjohnson** [comment]
This looks really good to me and a nice addition to the project!

**Comment by acjohnson** [comment]
Thanks @fsperling !

**Comment by ysde** [comment]
Hi @fsperling @acjohnson 
Thanks for the PR and the review

I think add `delete` command to the `cli` is fine, it's like a internal command.
Once the commands are enough for majority tasks, maybe we can extract it and make the `restore` and `backup` tasks as a wrapper on it.


> @fsperling I don't like the idea of adding bespoke arg parsers to this project, personally I think the `delete` command is perfectly acceptable to be added to `cli.py` maybe @ysde or others will have a different take.

Hi @fsperling  @acjohnson 
Thanks for the PR and review.
I also think it's ok to add to `cli.py`.

Thank you

**Comment by davidpellcb** [comment]
@acjohnson sorry to comment on an old PR but I came here looking for a way to turn off backing up dashboard `versions`. Is this possible? I just want the latest for my backups. Backing up the `versions` makes things take way too long for relatively minor benefit (in my case at least).

**Comment by fsperling** [comment]
@davidpellcb yes, when you create a backup you can specify the components you want. If you specify dashboards you only get the latest version of a dashboard.

components: folders,folder_permissions,dashboards,datasources,alert-channels,organizations,users,snapshots,versions,annotations

**Comment by davidpellcb** [comment]
Thanks @fsperling!

**Comment by davidpellcb** [comment]
@fsperling there seems to be a bug here when you remove anything from the `--components` list:

`KeyError: 'folder-permissions'`

If you don't pass `--components`, the code loops over the keys in the `backup_functions` dict: https://github.com/ysde/grafana-backup-tool/blob/42063c7773270a86d760fb60b377b7dbbea70d1a/grafana_backup/save.py#L51-L52

https://github.com/ysde/grafana-backup-tool/blob/42063c7773270a86d760fb60b377b7dbbea70d1a/grafana_backup/save.py#L22-L30

HOWEVER if you do pass `--components`, the code loops over everything in the list you passed and expects there to be a function of that name in the `backup_functions` list: https://github.com/ysde/grafana-backup-tool/blob/42063c7773270a86d760fb60b377b7dbbea70d1a/grafana_backup/save.py#L47-L48

...but there is no key `"folder-permissions"` in that dict.

**Comment by fsperling** [comment]
@davidpellcb Yes, I also noticed this error. We don't use folder-permissions so my workaround was to leave it out.
Please open a separate issue for this.

**Pull request merged**